libraries/lib-music-information-retrieval/MusicInformationRetrieval.cpp

```diff
@@ -9,35 +9,75 @@
 
 **********************************************************************/
 #include "MusicInformationRetrieval.h"
+#include "GetMeterUsingTatumQuantizationFit.h"
+#include "MirAudioReader.h"
+#include "MirUtils.h"
+#include "StftFrameProvider.h"
 
+#include <array>
 #include <cassert>
 #include <cmath>
+#include <numeric>
 #include <regex>
 
 namespace MIR
 {
 namespace
 {
+constexpr auto numTimeSignatures = static_cast<int>(TimeSignature::_count);
+
 auto RemovePathPrefix(const std::string& filename)
 {
    return filename.substr(filename.find_last_of("/\\") + 1);
 }
 
-// About right for a 4/4, which is the most common time signature.
-// When we get more clever, we may have different BPMs for different signatures,
-// i.e., about 64 BPM for 6/8, and 140 for 3/4.
-constexpr auto bpmExpectedValue = 120.;
-
 // When we get time-signature estimate, we may need a map for that, since 6/8
 // has 1.5 quarter notes per beat.
-constexpr auto quarternotesPerBeat = 1.;
+constexpr std::array<double, numTimeSignatures> quarternotesPerBeat { 2., 1.,
+                                                                      1., 1.5 };
+
+void FillMetadata(
+   const std::string& filename, const MirAudioReader& audio,
+   FalsePositiveTolerance tolerance,
+   const std::function<void(double progress)>& progressCallback,
+   std::optional<double>& bpm, std::optional<TimeSignature>& timeSignature)
+{
+   const auto bpmFromFilename = GetBpmFromFilename(filename);
+   if (bpmFromFilename.has_value())
+      bpm = bpmFromFilename;
+   else if (const auto meter =
+               GetMusicalMeterFromSignal(audio, tolerance, progressCallback);
+            meter.has_value())
+   {
+      bpm = meter->bpm;
+      timeSignature = meter->timeSignature;
```

Comment:
`timeSignature` is not assigned on all paths.  Better to make it consistent that way.

Or maybe better, why not return `MusicalMeter` by value and eliminate arguments?

Reply:
Absolutely.